README
PDFedit5 readme
==============

TOC
===
License
Contact
Prerequisites
	Base (kernel), pdfedit5-core-dev package
	Tests
	Gui
Documentation
Configuration
	Configure features
	Libraries and binaries specification
	Installation directories
Compilation
Installation
Cygwin build
FreeBSD build
NetBSD installation
Debian package
RPM based distros
Gentoo package
Altlinux 64b
PDFedit devel package


License
=======
PDFedit5 is distributed under terms of GNU GPL in version 2.
See doc/LICENSE.GPL for full license text.
See doc/AUTHORS for full list of authors and contributors.
For other more detailed documentation, look into "doc" subdirectory.
File doc/user/user_doc.html contain more detailed installation instructions
(in the Installation section) and list of required libraries.


Contact
=======
Feel free to contact me with any question via mail at info@danielripoll.me

Prerequisites
=============
Base (kernel), pdfedit5-code-dev package, xcb
---------------------------------------
Boost libraries must be installed. See detailed description about configure
parameters if you have some unusual installation.
Freetype library is almost necessary, because nearly all PDFs use these fonts.
T1 font library is also recommended.

Tests
-----
Unlike PDFedit, PDFEdit5 is still on development and and haven't been tested yet against CPPUNIT library,
so don't blame if something didn't work. Indeed, nothing work at this right moment.
See detailed description about configure parameters if you have some unusual installation.

Tools
-----
All tools use boost-program-options library so it has to be installed on your
system if you want to compile them (see later how to enable them). Some
distributions use a separate package (e.g. Debian) but most others include
it directly into boost package.
add_image tool uses libpng and its headers so libpng-dev package has to be 
installed.

Gui
---
Make sure you have all needed libraries (qt5, boost and xcb) along with their 
headers. 
You also need the environment variable QTDIR to be set to directory in which 
the Qt toolkit (version 5) is installed. Many distributions already set this 
when installing Qt, but some of them don't. In that case, you have to set the 
QTDIR manually. Usually it is set to something like /usr/lib/qt5 or /usr/share/qt5
(use e. g. export QTDIR=/usr/share/qt35to set the variable if this is your 
case).  
Also, while usually not needed, QMAKESPEC may be set to match your system type.
For most systems it is unnecessary to set it, although it may be needed if you
cross-compile (like producing i386 binaries on amd64 platform) or in some rare
configurations. Once you set QTDIR, you may try typing "ls $QTDIR/mkspecs".
QMAKESPEC should be set to name of one of the directories that are in mkspecs
subdirectory in the Qt directory.
See https://qt-project.org/wiki/Category:Tools or 
http://qt-project.org/doc/qt-5/qmake-manual.html for more information.
QTDIR is used also for qmake (tool for generating makefiles from QT project
files) detection. By default, this is stored somewhere under QTDIR directory
(typically QTDIR/bin/), but there are also platforms, where this is not 
exactly true. If you have such a platform, try to use --with-qmake-dir
parameter to configure (see bellow)

Documentation
=============
There are 3 levels of documentation in this project.
* Basic - in the form of the man page which is always installed 
* doxygen - programming documentation extracted from sources and
	    processed by doxygen tool. Obviously, doxygen has to
	    be installed.
	    This documentation is not created and installed by
	    default and you have to explicitly enable it by
	    --enable-doxygen-doc parameter to ./configure
	    Result documentation is placed in the DOC_PATH/doxygen
	    directory (DOC_PATH is PREFIX/share/doc/pdfedit)
	    It is strongly recommended for pdfedit-core-dev package
	    done by distributors.
* user manual - html pages with user manual about PDFedit usage.
	    This documentation is not created by default and you
	    have to enable it by configure parameter
	    --enable-user-manual. We are using xml based docbook
	    format, so you have to have docbook xsl and xsltproc
	    installed. 
	    Result documentation is placed in the DOC_PATH/manual
	    directory.
* advanced - design documentation which describes inner data 
	    structures, algorithms and relations from the 
	    programmer perspective. This one is useful if you
	    want to develop PDFedit5 or 3rd party application based
	    on PDFedit5.
	    This documentation is not created by default too and
	    you have to enable it by configure parameter 
	    --enable-advanced-doc. We are using xml based docbook
	    format, so you have to have docbook xsl and xsltproc
	    installed. 
	    Result documentation is placed in the DOC_PATH/design/
	    directory.

Configuration
=============
 Run:
  ./configure [parameters]

NOTE for those who use sources from GIT. configure is NO MORE part of the
GIT tree!!! You have to generate it. This can be done simply by
$ autoconf
in the root of the checkout project. Note that you have to have autoconf
package installed on your machine.

Simple run without any parameter results in default configuration where 
application is compiled in release mode (optimizations turned on, debug
symbols are not present, compilation warnings are eliminated), GUI is
enabled, doxygen documentation is generated, no kernel tests (stored in 
src/kernel/tests/) are compiled and application will be installed into the
/usr/local/pdfedit directory.

Configuration enables also compilation and installation in so called
pdfedit-core-dev package which prepares all libraries and header files for
3rd party application which want to reuse our core (kernel) functionality 
(see more in the Pdfedit devel package bellow).
This is not done by default and you have to enable it explicitly by
--enable-pdfedit-dev-core configure parameter.
Note that development package and gui can be mixed together but also single
dev. package can be created by disabling gui (see bellow).
pdfedit-dev-core installation comes with pdfedit-core-dev-config script
which can be used by 3rd party code for its configuration to correctly set
compiler flags include paths, libraries and so on.

Configure features
------------------
You can change this behavior with following parameters (feature is used
if --enable- prefix is used, contrary --disable- prefix turns off this
feature - e.g. if feature name stack-protector => --enable-stack-protector 
uses this feature):
	-stack-protector - adds stack protector parameter to the compilation
	 flags. Disabled by default, because not all compilers support this
	 feature (gcc > 4.1 supports it)

	-release - controls compilation flags for release mode (no debugging
	 needed). This is used by default. It means that optimizations are
	 turned on, debug information is not included in the result binary.
	 If you want to debug or do some development, --disable-release is
	 strongly recommended.

	-debug-info - if --enable-release is used and you still need debug
	 information (e.g. when application crashes to have some reasonable
	 stack traces) use --enable-debug-info. This parameter feature would
	 be ignored if --disable-release is used.
	
	-observer-debug - just for developers only. Disabled by default. If
	 turned on, some more debug information is added to the kernel code
	 to enable debugging observers based code.

	-gui - Creates GUI for PDFedit5 (pdfedit5 binary). Enabled by default. 
	 If --disable-gui is used, no GUI (no pdfedit binary is created).
	
	-kernel-tests - Compiles all kernel tests (in src/kernel/tests). 
	 Disabled by default and intended for developers/testers only. Note
	 that this requires to have CPPUNIT installed on your system.

	-tools - Compiles all tools which are available in src/tools.  Disabled
	 by default and intended for those interested in simple examples of
	 pdfedit-core-dev library usage and those who need some simple tools for
	 pdf manipulation/analyzes and don't want to use GUI or scripting
	 (pdf_to_text, flattener, delinerizator, pdf_object_comparer, etc.).
	 Note that you need also --enable-pdfedit-core-dev for tools.
	
	-doxygen-doc - generates also doxygen documentation useful for 
	 programmers. Disabled by default. Note that doxygen has to be
	 installed for this feature.

	-user-manual - generates also docbook user manual documentation.
	 Disabled by default. Note that docbook and its prerequisites
	 have to be installed for this feature.

	-advanced-doc - generates also docbook design documentation.
	 Disabled by default. Note that docbook and its prerequisites
	 have to be installed for this feature.
	
	-pdfedit-core-dev - generates libraries needed for 3rd party
	 applications which want to use pdfedit kernel core. Moreover 
	 make install will install also header files, libraries and
	 pdfedit-core-dev-config helper script (see Installation directories
	 bellow). This option is disabled by default.
	-poratability-flags - adds compiler flags which force c99/c++98 
	 standard and posix/ansi portability. This option is enabled by default.

Default configuration process will check for the number of CPUs/cores and 
use this value for make jobs parallelization (each make will use different
CPU/core). You can change this behavior with --with-parallel-make=VALUE 
where VALUE is one of the:
	- auto (default) - (on Linux - checks /proc/cpuinfo and use the CPU
	  counts for the number of parallel make jobs).
	- off - no parallelization
	- number - the number of parallel make jobs (positive number
	  expected)
Default behavior (no need for --with-parallel-make parameter) should be OK
in almost all situations (with benefit of shorter built time), however if 
you have specific requirements when you don't want/can't use all your
CPUs/cores for compilation you can use either off or precise number of
parallel jobs.


Libraries and binaries specification
------------------------------------
You can also control search paths for required libraries and binaries:

	- Freetype2 font library
	./configure --with-ft-prefix=PATH_TO_YOUR_FREETYPE2
	resp. (for T1)
	
	- T1 font library
	./configure --with-t1-includes=PATH_TO_YOUR_T1_HEADERS \
		--with-t1-library=PATH_TO_YOUR_T1_LIB
	
	- CPPUnit
	./configure  --with-cppunit-prefix=PATH_TO_YOUR_CPPUNIT

	- Boost
	./configure --with-boost=YOUR_BOOST_INSTALLATION_PATH \
		--with-boost-libdir=YOUR_BOOST_LIB_PATH 

	- qmake is searched in $QTDIR/bin, /usr/bin, /usr/local/bin,
	  /usr/lib/qt5/bin and in directories in $PATH. If you have
	  qmake installed in other directory, use:
	./configure --with-qmake=QMAKE_BINARY_WITH_PATH

	- QT comes with lrelease binary which is used for 
	  localization data files translation. This binary is
	  searched in $QTDIR/bin, /usr/bin, /usr/local/bin,
	  /usr/lib/qt5/bin and in directories in $PATH.
	  If it is not found on your system, please use
	./configure --with-lrelease-bin=LRELEASE_BINARY_WITH_PATH

	- doxygen is searched in /usr/bin , /usr/local/bin and 
	  current PATH directories. If you have doxygen installed
	  in other directory, use:
	./configure --with-doxygen=DOXYGEN_BINARY_WITH_PATH

	- xsltproc is  searched in /usr/bin , /usr/local/bin and 
	  current PATH directories. If you have xsltproc installed
	  in other directory, use:
	./configure --with-xsltproc=XSLTPROC_BINARY_WITH_PATH

	- docbook xslt files are necessary for xsltproc to generate
	  correct output from docbook xml files. There are many
	  places where these files can be found. As many systems
	  use different directories and there is no general way
	  to detect, we are using file doc/tools/docbook_xslt_paths
	  to define all possible directories which are searched.
	  If you are sure that you have installed package with
	  this file (usually stored ...html/dobook.xsl) add its 

Changelog
2014-04-20 PDF Editor 0.99
    *First commit, starting with Qt5, nothing is executable ATRM

2011-02-18 PDF Editor 0.4.5-79
	* Internal changes
		- string to font encoding fixed on the library level
		- C{Dict,Array,CStream}::getProperty<Type> added
		- C{Dict,Array} helpers cleanup
		- --enable-portable-flags configure option added
	* Bug 328, 355, 357, 359 fixed
	* Security fixes bacported from poppler (CVE-2010-3702, CVE-2010-4653
	  and CVE-2010-3704 received also CVE numbers)

2010-05-11 PDF Editor 0.4.5
	* fix for a bug caused by xpdf code refactoring (const methods and
	  parameters cleanup). 
	* Bugs 348, 350, 351 fixed

2010-04-13 PDF Editor 0.4.4.1
	* Fix compilation problems

2010-04-12 PDF Editor 0.4.4
	* config/pdfedit-core-dev.m4 added for easier pdfedit-core-dev
	  integration into 3rd party applications
	* xpdf code base updated to 3.02pl4 patch which fixes 
	  several serious security vulnerabilities
	* Bugs 320, 324, 329, 330, 334, 338, 343 fixed
	* several potential memory leaks fixed
	* Internal changes
		- CPdf::changeTrailer implemented (bt#224)
		- utils::isDescendant renamed to utils::isNodeDescendant
		- CPdf methods signiture cleanup - const and references for
		  parameters
		- --enable-tools configure option added and tools moved to src
		  (more in README)
	* Added scripting API to modify stream dictionaries
	* Raw object saving (thanks to Ron Kaminsky for patch)

2009-07-21 PDF Editor 0.4.3
	* xpdf code base updated to 3.02pl3 patch which fixes
	  several serious remote vulnerabilities
	* French translation update (bug 275)
	* Fix for [33853] Secunia advisory backported from poppler
	* Flattener class implemented (bt#289)
	* Bugs 248, 256, 285, 290, 291, 295, 299, 301, 304, 305, 306, 307, 309, 310, 314 fixed
	* Internal changes
		- all revisions are numbered in oposite direction (0 is 
		  always the oldest one) to enable stable numbering
		- changes are not discarded after we get to an older
		  revision and back

2008-12-15 PDF Editor 0.4.2
	* dependency to boost-iostreams dropped
	* Debian fixes (462544-crash, 444648-bad SCII85Encoding output) backported
	* changed streams are encoded with FlateDecode compression
	* read-only support for encrypted documents added
	* new --enable-pdfedit-core-dev parameter for configure for
	  development package installation (see README for more information)
	* Transformation matrix and page metrics can be changed for multiple
	  pages at once in GUI.
	* Configurable make parallelization added for more efficient building
	  See --with-parallel-make configure parameter (defaults to use all
	  available CPUs)
	* German localization updated by Andreas Mantke and Georg Hofmann
	* Chinese (Traditional) localization added - thanks to Yi-Shun Wang
	* French localization added - thanks to Sylvain
	* Bugs 131, 157, 158, 195, 199, 207, 215, 217, 220, 222, 217, 226, 
	  228, 233, 241, 246, 247, 249, 257, 258, 260, 262, 267, 270, 272 fixed
	* Print summary information with used options after running configure script
	* 64b distributions with boost libraries in lib64 directory can use
	  --with-boost-libdir configure parameter
	* Internal changes
		- versioning updated - released and dirty or CVS builds cotain
		  the timestamp of the most recently changed file
		- Removed unused obsolete code (QSA for Qt4) from source tree
		- smart pointers are used also for CPdf instance
		- support for stream encoding in the pdfwriter
		- JBIG2Stream::clone implemented
	
2008-02-24 PDF Editor 0.4.1
	* Missing src/kernel/Makefile in source tarball fixed
	* Fixed bug in pdfedit.pro which could result in pdfedit binary
	  not being installed

2008-02-22 PDF Editor 0.4.0
	* some memory leaks fixed - thanks to Gwurk
	* new XPDF 3.02 imported to the tree - this fixed some security
	  vulnerabilities reported by secunia as well as provide with 
	  the new functionality (support for PDF 1.6 and 1.7). 
	  See src/xpdf/CHANGES for more information
	* Internal: 
		- debuging for observers, more effective CObject -> Object
		  translation for CDict and CArray
		- new makefiles infrastructure (see README for more information).
		  WARNING:
		  For those who get sources directly from CVS, note that you
		  have to generate configure script by hand, because it is 
		  no more a part of CVS tree (simply run autoconf without
		  any parameters).
		- delinearizator moved from src/utils/ to src/kernel
		- Improved testing environment
	* Code cleanups by Gwurk
	* Bugs 123, 137, 144, 147, 152, 153, 154, 155, 156, 162, 164, 165,
	  168, 169, 175, 178, 179, 180, 182, 183, 187, 201, 202, 203, 204,
	  206, 208 fixed
	* Added Tango icon theme from Tiziano Mattei.
	* Fixed bug with operator-only command available for non-operators too,
	  leading to errors
	* kernel+xpdf+utils+tests ported to Windows (vc2005 project)
	* Fixed observer unregistering

2007-08-28 PDF Editor 0.3.2
	* bug #160 (http://pdfedit.petricek.net/bt/view.php?id=160) fixed
	* README updated for several platforms installation description
	* Fixed bug in Xpdf StreamPredictor announced by http://secunia.com/advisories/26188/
	* Fixed bug when doxygen would overwrite /dev/null device if run under root
	  (thanks to Bernard Johnson for reporting)
	* Better error message when invalid command line option is specified
	* QMAKESPEC is not required by configure
	* Minor fixes in documentation
	* Fixed bug in script used to make tarballs - one documentation file
	  was not included in releases
	* Fixed bad format string error on 64bit platforms

2007-04-29 PDF Editor 0.3.1
	* Fixed command line function invocation
	* Fixed QSA using bad qmake
	* Fixed --enable-release option in configure script
	* Xpdf code cleanup (removed unused files)
	* Fixed possible unicode string handling errors
	* Added checks for QTDIR and QMAKESPEC to ./configure
	* Compilation fixes for gcc-4.3 (thanks to Martin Michlmayr for patch)
	* Added German translation (thanks to Andreas Mantke)

src/gui/util.cc
/*
 * PDFedit - free program for PDF document manipulation.
 * Copyright (C) 2006-2009  PDFedit team: Michal Hocko,
 *                                        Jozef Misutka,
 *                                        Martin Petricek
 *                   Former team members: Miroslav Jahoda
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program (in doc/LICENSE.GPL); if not, write to the 
 * Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, 
 * MA  02111-1307  USA
 *
 * Project is hosted on http://sourceforge.net/projects/pdfedit
 */
/** @file
 Various utility functions (string processing, file loading, debugging, etc ...)
 @author Martin Petricek
 \brief Utility functions
*/
#include "util.h"
#include "qtcompat.h"
#include <iostream>
#include <assert.h>
#include QBYTEARRAY
#include <QtGui/QColor>
#include <QtCore/QFile>
#include <QtCore/QObject>
#include <QtCore/QRegExp>
#include <QtCore/QString>
#include <QtCore/QStringList>
#include <QtCore/QTextCodec>
#include <QtCore/QTextSteam>
#include <QtCore/QVariant>
#include <QtWidgets/QWidget>
#include <kernel/static.h>
#include <utils/debug.h>
#include QLIST

namespace util {

using namespace std;

namespace {

	//
	// Pdfedit encoding class
	//
	class PdfeditEncoding; //forward declaration
	/** Single instance of PdfeditEncoding class */
	PdfeditEncoding *encoding_instance=0;
	/**
	 * This class is used because of special pdf accented character encoding.
	 */
	class PdfeditEncoding : public QTextCodec
	{
		typedef std::map<unsigned short, string> UMap;

		// Display this char when mapping is not known
		const static char UNKNOWN_CHAR = '?';

		private:
			UMap umap;	// Mapping from unicode to windows character encoding

			//
			// Ctor - singleton
			//
		private:
			PdfeditEncoding ()
			{
				umap[271] = string (1,'d') + string (1,(char)39);	// daccent
				umap[318] = string (1,'l') + string (1,(char)39);	// laccent
				umap[352] = string (1,(char)138);					// Scaron
				umap[353] = string (1,(char)154);					// scaron
				umap[357] = string (1,'t') + string (1,(char)39);	// tcaron
				umap[381] = string (1,(char)142);					// Zcaron
				umap[382] = string (1,(char)158);					// zcaron
			
			};

			//
			// Instance
			//
		public:
			static PdfeditEncoding* get ()
			{
				if (!encoding_instance) encoding_instance=new PdfeditEncoding();
				return encoding_instance;
			}

			//
			// Interface
			//
		public:
#ifdef QT4
			virtual QByteArray name () const
#else
			virtual const char* name () const
#endif
			{
				static const char* _name = "Pdfedit accented windows encoding";
				return _name;
			}
			int mibEnum () const
				{ return 2252; /* windows 1252 */};
			int heuristicContentMatch (const char* ,int) const
				{ return 0; }


			QString
			toUnicode (const char* chars, int len ) const
			{
				return QString::fromLatin1(string(chars,len).c_str());
			}
#ifdef QT4
			QString convertToUnicode (const char* chars, int len, QTextCodec::ConverterState* ) const
			{
				return toUnicode(chars,len);
			}
			QByteArray convertFromUnicode (const QChar* chars, int len, QTextCodec::ConverterState* ) const
			{
				string tmp;
				for (int i = 0; i < len; ++i)
				{
					QChar c =chars[i];
					UMap::const_iterator it = umap.find (c.unicode());
					if (it != umap.end())
						tmp += (*it).second;
					else
						tmp += c.latin1() ? c.latin1() : UNKNOWN_CHAR ;
				}
				assert (tmp.size() >= len);
				return QCString (tmp.c_str());
			}
#else
			QCString
			fromUnicode (const QString& uc, int& lenInOut) const
			{
				string tmp;
				for (size_t i = 0; i < uc.length(); ++i)
				{
					QChar c (uc[i]);
					UMap::const_iterator it = umap.find (c.unicode());
					if (it != umap.end())
						tmp += (*it).second;
					else
						tmp += c.latin1() ? c.latin1() : UNKNOWN_CHAR ;
				}
				lenInOut = tmp.size();
				assert (tmp.size() >= uc.length());
				return QCString (tmp.c_str());
			}
#endif
	
	}; // class PdfeditEncoding

} // namespace

/**
 Convert array of double values to variant
 @param d Pointer to array fo doubles
 @param count number of elements in the array
*/
QVariant varFromDoubleArray(double *d,int count) {
 Q_List<QVariant> rPos;
 for (int i=0;i<count;i++) {
  rPos.append(d[i]);
 }
 return QVariant(rPos);
}

/**
 Convert variant to array of doubles. If size of variant array is larger than out_size,
 only part will be converted. If size is smaller, output array will be zero-padded
 @param v Variant to convert
 @param out pointer to array of doubles
 @param out_size number of elements in output array.
 @return number of elements converted (between 0 and out_size)
*/
int varToDoubleArray(const QVariant &v,double *out,int out_size) {
 Q_List<QVariant> list=v.toList();
 Q_List<QVariant>::const_iterator it=list.constBegin();
 int al=0;
 while(it!=list.constEnd()) {
  if (al>=out_size) break;//We filled all values
  out[al]=(*it).toDouble();
  it++;
  al++;
 }
 //Pad with zeroes
 int alx=al;
 while (alx<out_size) {
  out[alx]=0;
  alx++;
 }
 return al;
}

/**
 Prints error message and terminates application
 @param message Error message to show
*/
void fatalError(const QString &message){
 cerr << endl
      << convertFromUnicode(QObject::tr("Fatal Error"),CON) << "!" << endl
      << convertFromUnicode(message,CON) << endl;
 exit(-1);
}

/**
 Convert < > and & characters to corresponding html entities (&gt; &lt; &amp;)
 @param str String to convert
 @return Converted string
*/
QString htmlEnt(const QString &str) {
 QString res=str;
 res.replace('&',"&amp;");
 res.replace('>',"&gt;");
 res.replace('<',"&lt;");
 return res;
}

/**
 Find first occurence of separator in string and return its position, ignoring backslash-escaped separators
 If separator is not found, -1 is returned
 @param start Start searching from this position in string
 @param separator Separator of string elements
 @param line String to look for separator
 @return position of first separator from start position, ignoring backslash-escaped separators
*/
int findSep(int start,const QString &line,char separator) {
 int pos=line.find(separator,start);
 while(true) {
  if (pos==-1) {
   return -1;
  } else { //Found
   if (pos>start) { //Want escaped character and this is not first character
    if (line.at(pos-1)=='\\') { //Preceding character is a backslash -> need to track back
     int count=0;
     while(pos>count && line.at(pos-count-1)=='\\') {
      count++;
     }
     if (!(count&1)) {
      // Even number of backslashes -> not escaped character, only escaped backslash
      return pos;
     }
     pos=line.find(separator,pos+1);//Look for next character
     continue;//Restart loop
    }
   }
   return pos;
  }
 }
}

/**
 splits QString containing elements separated with given character
 All whitespaces from beginning and end of elements are trimmed
 @param separator Separator of elements
 @param line String containing elements separated with separator
 @param escape If true, "\\" will be converted to "\" and \(separator) will be converted to (separator) and not treated as separator
 @return QStringlist with elements
*/
QStringList explode(char separator,const QString &line,bool escape/*=false*/) {
 QStringList qs;
 if (escape) {
  int pos=0;
  QString add;
  while(true) {
   int nPos=findSep(pos,line,separator);
   if (nPos<0) {
    add=line.mid(pos);//rest of string
   } else {
    add=line.mid(pos,nPos-pos);//Part of string
   }
   escapeSlash(add);
   qs+=add;
   if (nPos<0) break;
   pos=nPos+1;//Skip separatr and look for next
  }
 } else {
  qs=QStringList::split(separator,line,TRUE);
 }
 for (unsigned int i=0;i<qs.count();i++) {
  qs[i]=qs[i].stripWhiteSpace();
 }
 return qs;
}

/**
 Load content of file to string.
 NULL string is returned if file does not exist or is unreadable.
 It is assumed that the file is in utf8 encoding
 @param name Filename of file to load
 @return file contents in string.
*/
QString loadFromFile(const QString &name) {
 QFile f(name);
 if (!f.open(IO_ReadOnly)) {
  //Failure
  return QString::null;
 }
 QByteArray qb=f.readAll();
 f.close();
 QTextCodec *codec=QTextCodec::codecForName("utf8");
 assert(codec);
 QString res=codec->toUnicode(qb);
 return res;
}

/**
 Save string into file with utf8 encoding.
 @param name Filename of file to write to. It will be overwritten.
 @param content String to write.
 @return true in case of success, false in case of failure
*/
bool saveToFile(const QString &name,const QString &content) {
 QFile f(name);
 if (!f.open(IO_WriteOnly)) {
  //Failure
  return false;
 }
 QTextCodec *codec=QTextCodec::codecForName("utf8");
 assert(codec);
 QCString qb=codec->fromUnicode(content);
 f.writeBlock(qb.data(),qb.length());
 f.close();

 return true;
}

/**
 Save raw bytes into file.
 @param name Filename of file to write to. It will be overwritten.
 @param content Buffer to write.
 @return true in case of success, false in case of failure
*/
bool saveRawToFile(const QString &name, const QByteArray &content) {
 QFile f(name);
 if (!f.open(IO_WriteOnly)) {
  //Failure
  return false;
 }

 f.writeBlock(content.data(),content.size());

 f.close();
 return true;
}

/**
 Print stringlist to stdout
 @param l String list to print
*/
void printList(const QStringList &l) {
 QStringList::ConstIterator it=l.begin();
 for (;it!=l.end();++it) { //load all subitems
  QString x=*it;
  cout << Q_OUT(x) <<endl;
 }
}

/**
 Write line to specified logfile
 @param message Line to write to logfile
 @param fileName Name of log file. If this is null or empty string, nothing is done
*/
void consoleLog(const QString &message,const QString &fileName) {
 if (fileName.isNull()) return;
 if (fileName=="") return;
 QFile con(fileName);
 con.open(IO_WriteOnly | IO_Append);
 QTextStream conOut(&con);
 conOut << message << "\n";
 con.close();
}

/**
 Replace two backslashes followed by any character (escaped character) with that character
 @param line String in which to remove extra backslashes (in,out)
*/
void escapeSlash(QString &line) {
 static QRegExp slash("\\\\(.)");
 line.replace(slash,"\\1");
}

/**
 Return string from line up to first separator character and remove that string and the separator from the line.
 If separator is not found, entire string is returned and the line is set to empty string
 @param separator Separator of string elements
 @param line line to get from and remove first element
 @param escape If true, "\\" will be converted to "\" and \(separator) will be converted to (separator) and not treated as separator
 @return first string element (contents of string until separator)
*/
QString getUntil(char separator,QString &line,bool escape/*=false*/) {
 int pos=line.find(separator);
 while(true) {
  if (pos==-1) { //Not found
   QString _line=line;
   line="";
   if (escape) escapeSlash(_line);
   return _line;
  } else { //Found
   if (escape && pos>0) { //Want escaped character and this is not first character
    if (line.at(pos-1)=='\\') { //Preceding character is a backslash
     pos=line.find(separator,pos+1);//Look for next character
     continue;//Restart loop
    }
   }
   QString first=line.left(pos);
   line=line.mid(pos+1);
   if (escape) escapeSlash(first);
   return first;
  }
 }
}

/**
 Set debugging verbosity level. Accept debugging level either as number, or as one of symbolic constants:<br>
 PANIC, CRITICAL, ERROR, WARNING, INFO, DEBUG, DBG (=DEBUG)<br>
 Constants are case insensitive, starting part of constant is sufficient<br>
 Only messages with priority equal or higher to specified level are shown
 @param param new debugging level
*/
void setDebugLevel(const QString &param){
 QString cns=param.upper();
 if (cns.length()) { //Check for symbolic constants
  if (QString("PANIC").startsWith(cns))		{ debug::changeDebugLevel(debug::DBG_PANIC); return; }
  if (QString("CRITICAL").startsWith(cns))	{ debug::changeDebugLevel(debug::DBG_CRIT); return; }
  if (QString("ERROR").startsWith(cns))		{ debug::changeDebugLevel(debug::DBG_ERR); return; }
  if (QString("WARNING").startsWith(cns))	{ debug::changeDebugLevel(debug::DBG_WARN); return; }
  if (QString("INFO").startsWith(cns))		{ debug::changeDebugLevel(debug::DBG_INFO); return; }
  if (QString("DEBUG").startsWith(cns))		{ debug::changeDebugLevel(debug::DBG_DBG); return; }
  if (QString("DBG").startsWith(cns))		{ debug::changeDebugLevel(debug::DBG_DBG); return; }
 }
 //If debuglevel is set outside of limits - no problem, nearest "in limits" value is defacto used
 debug::changeDebugLevel(atoi(param));
 //If non-number is given, default 0 is silently used ... :)
}

/**
 Return correctly localized string telling count of some items.
 Some languages (for example Czech) have more plural forms (2-4 items / 5 or more items), which is handled here
 @param count Count of items
 @param singular English signular form of the noun, without space before the word
 @param plural English plural form of the noun
 @return Localized string
*/
QString countString(int count,QString singular,QString plural) {
 QString str=QString::number(count)+" ";
 if (count==1) str+=QObject::tr(singular,"1");
 else if (count>=2 && count<=4) str+=QObject::tr(plural,"2-4");
 else str+=QObject::tr(plural,"5+");
 return str;
}

/**
 Mix colors from oldColor and newColor according to weight
 Weight should be between 0 and 1,
 1 meaning the color will be "new",
 0 meaning the color will be "old",
 everything else mean linear interpolation between new color (with given weight) and old color (with weight of (1-weight))
 @param weight Weight of new color.
 @param oldColor Old color
 @param newColor New color
*/
QColor mixColor(const QColor &oldColor,double weight,const QColor &newColor) {
 QColor retCol(
  (int)(oldColor.red()*(1-weight)+newColor.red()*weight),
  (int)(oldColor.green()*(1-weight)+newColor.green()*weight),
  (int)(oldColor.blue()*(1-weight)+newColor.blue()*weight)
 );
 return retCol;
}

/**
 Modify widget foreground and background color by blending
 with new colors using given weight
 @param widget Widget to modify
 @param fg new foreground color
 @param weight_fg weight of new foreground color
 @param bg new background color
 @param weight_bg weight of new background color
*/
void colorMod(QWidget* widget,QColor fg,double weight_fg,QColor bg,double weight_bg) {
 if (weight_bg>0) {
  QColor oldColorBack=widget->paletteBackgroundColor();
  QColor newColorBack=mixColor(oldColorBack,weight_bg,bg);
  widget->setPaletteBackgroundColor(newColorBack);
 }
 if (weight_fg>0) {
  QColor oldColorFore=widget->paletteForegroundColor();
  QColor newColorFore=mixColor(oldColorFore,weight_fg,fg);
  widget->setPaletteForegroundColor(newColorFore);
 }
 //TODO: check contrast of new color and modify them if necessary
}

/**
 Return string list containing 'count' items, numbered from 'start' to 'start+count-1'
 @param count number of items in list
 @param start first item in the list
*/
QStringList countList(unsigned int count,unsigned int start/*=0*/) {
 QStringList items;
 for(unsigned int i=0;i<count;i++) {
  items+=QString::number(i+start);
 }
 return items;
}

/**
 Return list of supported text encodings
 (Based on support from QTextCodec in Qt)
*/
QStringList supportedEncodings() {
 return QStringList::split(",",",Apple Roman,Big5,Big5-HKSCS,\
CP874,CP1250,CP1251,CP1252,CP1253,CP1254,CP1255,CP1256,CP1257,CP1258,\
GB2312,GB18030,GBK,IBM 850,IBM 866,\
ISO8859-1,ISO8859-2,ISO8859-3,ISO8859-4,ISO8859-5,ISO8859-6,ISO8859-7,ISO8859-8,ISO8859-8-i,ISO8859-9,\
ISO8859-10,ISO8859-13,ISO8859-14,ISO8859-15,\
JIS7,KOI8-R,KOI8-U,Latin1,Shift-JIS,TIS-620TSCII,eucJP,eucKR,utf8,utf16," + QString(PdfeditEncoding::get()->name()),true);
}

/** Codec for default charset */
QTextCodec *defCodec=NULL;

/**
 set default charset used when reading (and writing) property values
 @param charsetName charset name used for all 8bit to/from unicode conversions
*/
void setDefaultCharset(const QString &charsetName) {
 //Codecs should not be deleted, as Qt is managing them
 if (defCodec) defCodec=NULL;
 // Is it our codec?
 if (PdfeditEncoding::get()->name() == charsetName)
 {	
 	defCodec = PdfeditEncoding::get();
	return;
 }
 // Default codecs
 if (charsetName.isNull() || charsetName=="") return;//no codec
 defCodec=QTextCodec::codecForName(charsetName);
}

/**
 Return QString from std::string with optional character conversion according to settings
 @param str String to convert
 @param ctx Context of conversion (PDF, console, filename ...)
 @return converted string
 */
QString convertToUnicode(const std::string &str, CharsetContext ctx) {
 /** \todo: handle context, different settings for different contexts */
 if (ctx==UTF8) {
  return QString::fromUtf8(str.c_str());
 }
 if (ctx==CON) {
  return QString::fromLocal8Bit(str.c_str());
 }
 if (ctx==NAME) {
  return QFile::decodeName(str.c_str());
 }
 //PDF
 if (defCodec) {
  return defCodec->toUnicode(str.c_str());
 }
 return QString::fromLatin1(str.c_str());
}

/**
 Return std::string from QString with optional character conversion according to settings
 @param str String to convert
 @param ctx Context of conversion (PDF, console, filename ...)
 @return converted string
 */
std::string convertFromUnicode(const QString &str, CharsetContext ctx) {
 /** \todo: handle context, different settings for different contexts */
 if (ctx==UTF8) {
  return std::string(str.utf8());
 }
 if (ctx==CON) {
  return std::string(str.local8Bit());
 }
 if (ctx==NAME) {
  return std::string(QFile::encodeName(str).data());
 }
 //PDF
 if (defCodec) {
  QCString cstr=defCodec->fromUnicode(str);
  return std::string(cstr);
 }
 //Fallback to Latin1 if no codec is set
 return std::string(str.latin1());
}

} //namespace util
